File: sdks/rs/tests/dsn_test.rs
// Shared test vectors from loza/spec/dsn/test-cases.json (25 cases: 12 valid, 13 invalid).
type ExpectedDsn<'a> = (
    &'a str,
    &'a str,
    u16,
    &'a str,
    &'a str,
    &'a str,
    bool,
    &'a str,
    &'a str,
    Option<&'a str>,
    Option<&'a str>,
    Option<&'a str>,
    Option<&'a str>,
);

fn assert_valid(input: &str, expected: ExpectedDsn<'_>) {
    let (
        expect_scheme,
        expect_host,
        expect_port,
        expect_project,
        expect_env,
        expect_service,
        expect_tls,
        expect_transport,
        expect_base_url,
        expect_events_url,
        expect_batch_url,
        expect_otlp_url,
        expect_tail_ws_url,
    ) = expected;
    let dsn = loza::dsn::parse(input).unwrap_or_else(|_| panic!("expected valid DSN: {input}"));
    assert_eq!(dsn.scheme, expect_scheme, "scheme mismatch for {input}");
    assert_eq!(dsn.host, expect_host, "host mismatch for {input}");
    assert_eq!(dsn.port, expect_port, "port mismatch for {input}");
    assert_eq!(dsn.project, expect_project, "project mismatch for {input}");
    assert_eq!(
        dsn.collector_name, expect_project,
        "collector_name mismatch for {input}"
    );
    assert_eq!(dsn.env, expect_env, "env mismatch for {input}");
    assert_eq!(dsn.service, expect_service, "service mismatch for {input}");
    assert_eq!(dsn.tls, expect_tls, "tls mismatch for {input}");
    assert_eq!(
        dsn.transport, expect_transport,
        "transport mismatch for {input}"
    );
    assert_eq!(
        dsn.base_url, expect_base_url,
        "base_url mismatch for {input}"
    );
    if let Some(ev) = expect_events_url {
        assert_eq!(dsn.events_url, ev, "events_url mismatch for {input}");
    }
    if let Some(b) = expect_batch_url {
        assert_eq!(dsn.batch_url, b, "batch_url mismatch for {input}");
    }
    if let Some(o) = expect_otlp_url {
        assert_eq!(dsn.otlp_url, o, "otlp_url mismatch for {input}");
    }
    if let Some(t) = expect_tail_ws_url {
        assert_eq!(dsn.tail_ws_url, t, "tail_ws_url mismatch for {input}");
    }
}

fn assert_invalid(input: &str) {
    let result = loza::dsn::parse(input);
    assert!(result.is_err(), "expected invalid DSN but got: {result:?}");
}

// ── Valid cases (12) ────────────────────────────────────────────────────────

#[test]
fn localhost_dev_with_explicit_port() {
    assert_valid(
        "loza://localhost:9308/demo?tls=false",
        (
            "loza",
            "localhost",
            9308,
            "demo",
            "default",
            "",
            false,
            "http",
            "http://localhost:9308",
            Some("http://localhost:9308/collectors/demo/events"),
            Some("http://localhost:9308/collectors/demo/events/batch"),
            Some("http://localhost:9308/collectors/demo/otlp/logs"),
            Some("ws://localhost:9308/collectors/demo/tail"),
        ),
    );
}

#[test]
fn localhost_default_port_9308() {
    assert_valid(
        "loza://localhost/demo?tls=false",
        (
            "loza",
            "localhost",
            9308,
            "demo",
            "default",
            "",
            false,
            "http",
            "http://localhost:9308",
            Some("http://localhost:9308/collectors/demo/events"),
            Some("http://localhost:9308/collectors/demo/events/batch"),
            Some("http://localhost:9308/collectors/demo/otlp/logs"),
            Some("ws://localhost:9308/collectors/demo/tail"),
        ),
    );
}

#[test]
fn prod_default_tls_true() {
    assert_valid(
        "loza://collector.example.com/demo",
        (
            "loza",
            "collector.example.com",
            443,
            "demo",
            "default",
            "",
            true,
            "http",
            "https://collector.example.com:443",
            Some("https://collector.example.com:443/collectors/demo/events"),
            Some("https://collector.example.com:443/collectors/demo/events/batch"),
            Some("https://collector.example.com:443/collectors/demo/otlp/logs"),
            Some("wss://collector.example.com:443/collectors/demo/tail"),
        ),
    );
}

#[test]
fn custom_env_and_service() {
    assert_valid(
        "loza://collector.example.com/demo?env=prod&service=api",
        (
            "loza",
            "collector.example.com",
            443,
            "demo",
            "prod",
            "api",
            true,
            "http",
            "https://collector.example.com:443",
            None,
            None,
            None,
            None,
        ),
    );
}

#[test]
fn otlp_transport() {
    assert_valid(
        "loza://collector.example.com/demo?transport=otlp",
        (
            "loza",
            "collector.example.com",
            443,
            "demo",
            "default",
            "",
            true,
            "otlp",
            "https://collector.example.com:443",
            None,
            None,
            Some("https://collector.example.com:443/collectors/demo/otlp/logs"),
            None,
        ),
    );
}

#[test]
fn grpc_transport() {
    assert_valid(
        "loza://collector.example.com/demo?transport=grpc",
        (
            "loza",
            "collector.example.com",
            443,
            "demo",
            "default",
            "",
            true,
            "grpc",
            "https://collector.example.com:443",
            None,
            None,
            None,
            None,
        ),
    );
}

#[test]
fn loopback_127_defaults_tls_false() {
    assert_valid(
        "loza://127.0.0.1/demo",
        (
            "loza",
            "127.0.0.1",
            9308,
            "demo",
            "default",
            "",
            false,
            "http",
            "http://127.0.0.1:9308",
            None,
            None,
            None,
            None,
        ),
    );
}

#[test]
fn ipv6_loopback_defaults_tls_false() {
    assert_valid(
        "loza://[::1]/demo",
        (
            "loza",
            "::1",
            9308,
            "demo",
            "default",
            "",
            false,
            "http",
            "http://[::1]:9308",
            None,
            None,
            None,
            None,
        ),
    );
}

#[test]
fn tls_auto_keeps_localhost_default() {
    assert_valid(
        "loza://localhost/demo?tls=auto",
        (
            "loza",
            "localhost",
            9308,
            "demo",
            "default",
            "",
            false,
            "http",
            "http://localhost:9308",
            None,
            None,
            None,
            None,
        ),
    );
}

#[test]
fn tls_auto_keeps_remote_default() {
    assert_valid(
        "loza://collector.example.com/demo?tls=auto",
        (
            "loza",
            "collector.example.com",
            443,
            "demo",
            "default",
            "",
            true,
            "http",
            "https://collector.example.com:443",
            None,
            None,
            None,
            None,
        ),
    );
}

#[test]
fn explicit_tls_true_on_localhost() {
    assert_valid(
        "loza://localhost:8443/demo?tls=true",
        (
            "loza",
            "localhost",
            8443,
            "demo",
            "default",
            "",
            true,
            "http",
            "https://localhost:8443",
            None,
            None,
            None,
            None,
        ),
    );
}

#[test]
fn explicit_port_4318_with_otlp() {
    assert_valid(
        "loza://collector.example.com:4318/backend?env=staging&service=auth&transport=otlp",
        (
            "loza",
            "collector.example.com",
            4318,
            "backend",
            "staging",
            "auth",
            true,
            "otlp",
            "https://collector.example.com:4318",
            None,
            None,
            None,
            None,
        ),
    );
}

// ── Invalid cases (13) ──────────────────────────────────────────────────────

#[test]
fn reject_empty_string() {
    assert_invalid("");
}

#[test]
fn reject_wrong_scheme_https() {
    assert_invalid("https://collector.example.com/demo");
}

#[test]
fn reject_wrong_scheme_http() {
    assert_invalid("http://collector.example.com/demo");
}

#[test]
fn reject_no_host() {
    assert_invalid("loza://");
}

#[test]
fn reject_triple_slash_no_host() {
    assert_invalid("loza:///project");
}

#[test]
fn reject_no_project() {
    assert_invalid("loza://collector.example.com");
}

#[test]
fn reject_empty_project() {
    assert_invalid("loza://collector.example.com/");
}

#[test]
fn reject_userinfo_key() {
    assert_invalid("loza://key@collector.example.com/demo");
}

#[test]
fn parse_userinfo_credentials() {
    let dsn = loza::dsn::parse("loza://user:pass@collector.example.com/demo")
        .expect("credentialed DSN should parse");
    assert_eq!(dsn.username.as_deref(), Some("user"));
    assert_eq!(dsn.password.as_deref(), Some("pass"));
    assert!(!dsn.base_url.contains("user"));
    assert!(!dsn.base_url.contains("pass"));
    assert!(!format!("{dsn:?}").contains("Some(\"pass\")"));
}

#[test]
fn public_credentials_route_and_redact_bearer_capability() {
    let capability = "lz_pub_6DJvd3D0izOaQx3n5BhKqN";
    let dsn = loza::dsn::parse(&format!(
        "loza://{capability}:@collector.example.com/public-collector"
    ))
    .expect("public DSN must parse");

    assert_eq!(dsn.username.as_deref(), Some(capability));
    assert_eq!(dsn.password.as_deref(), Some(""));
    assert_eq!(dsn.collector_name, "public-collector");
    assert_eq!(
        dsn.events_url,
        "https://collector.example.com:443/collectors/public-collector/events"
    );
    assert!(!format!("{dsn:?}").contains(capability));
}

#[test]
fn reject_invalid_tls_value() {
    assert_invalid("loza://collector.example.com/demo?tls=maybe");
}

#[test]
fn reject_invalid_transport_value() {
    assert_invalid("loza://collector.example.com/demo?transport=random");
}

#[test]
fn reject_port_zero() {
    assert_invalid("loza://collector.example.com:0/demo");
}

#[test]
fn reject_port_above_65535() {
    assert_invalid("loza://collector.example.com:99999/demo");
}

#[test]
fn reject_non_numeric_port() {
    assert_invalid("loza://collector.example.com:abc/demo");
}

#[test]
fn scoped_lql_query_url_is_canonical() {
    let dsn = loza::dsn::parse("loza://collector.example.com/demo").unwrap();
    assert_eq!(
        dsn.lql_url,
        "https://collector.example.com:443/collectors/demo/lql/query"
    );
    assert_eq!(dsn.lql_query_url, dsn.lql_url);
}

File: spec/dsn/dsn_test.go
package dsn

import (
	"fmt"
	"strings"
	"testing"
)

func TestParse(t *testing.T) {
	tests := []struct {
		name      string
		input     string
		valid     bool
		username  string
		password  string
		host      string
		port      int
		project   string
		env       string
		service   string
		tls       bool
		transport string
		baseURL   string
		eventsURL string
		batchURL  string
		otlpURL   string
		tailWSURL string
	}{
		// ── Valid cases ───────────────────────────────────────────────────────
		{
			name:      "localhost dev with explicit port and tls=false",
			input:     "loza://localhost:9308/demo?tls=false",
			valid:     true,
			host:      "localhost",
			port:      9308,
			project:   "demo",
			env:       "default",
			tls:       false,
			transport: "http",
			baseURL:   "http://localhost:9308",
			eventsURL: "http://localhost:9308/events",
			batchURL:  "http://localhost:9308/events/batch",
			otlpURL:   "http://localhost:9308/otlp/logs",
			tailWSURL: "ws://localhost:9308/tail",
		},
		{
			name:      "localhost default port 9308",
			input:     "loza://localhost/demo?tls=false",
			valid:     true,
			host:      "localhost",
			port:      9308,
			project:   "demo",
			tls:       false,
			transport: "http",
			baseURL:   "http://localhost:9308",
			eventsURL: "http://localhost:9308/events",
			tailWSURL: "ws://localhost:9308/tail",
		},
		{
			name:      "prod default tls=true",
			input:     "loza://collector.example.com/demo",
			valid:     true,
			host:      "collector.example.com",
			port:      443,
			project:   "demo",
			env:       "default",
			tls:       true,
			transport: "http",
			baseURL:   "https://collector.example.com:443",
			eventsURL: "https://collector.example.com:443/events",
			batchURL:  "https://collector.example.com:443/events/batch",
			otlpURL:   "https://collector.example.com:443/otlp/logs",
			tailWSURL: "wss://collector.example.com:443/tail",
		},
		{
			name:      "custom env and service",
			input:     "loza://collector.example.com/demo?env=prod&service=api",
			valid:     true,
			host:      "collector.example.com",
			port:      443,
			project:   "demo",
			env:       "prod",
			service:   "api",
			tls:       true,
			transport: "http",
			baseURL:   "https://collector.example.com:443",
		},
		{
			name:      "otlp transport",
			input:     "loza://collector.example.com/demo?transport=otlp",
			valid:     true,
			host:      "collector.example.com",
			port:      443,
			project:   "demo",
			tls:       true,
			transport: "otlp",
			baseURL:   "https://collector.example.com:443",
		},
		{
			name:      "grpc transport",
			input:     "loza://collector.example.com/demo?transport=grpc",
			valid:     true,
			host:      "collector.example.com",
			port:      443,
			project:   "demo",
			tls:       true,
			transport: "grpc",
			baseURL:   "https://collector.example.com:443",
			eventsURL: "https://collector.example.com:443/events",
			batchURL:  "https://collector.example.com:443/events/batch",
			otlpURL:   "https://collector.example.com:443/otlp/logs",
			tailWSURL: "wss://collector.example.com:443/tail",
		},
		{
			name:    "127.0.0.1 defaults to tls=false and port 9308",
			input:   "loza://127.0.0.1/demo",
			valid:   true,
			host:    "127.0.0.1",
			port:    9308,
			project: "demo",
			tls:     false,
			baseURL: "http://127.0.0.1:9308",
		},
		{
			name:    "::1 defaults to tls=false and port 9308 (brackets required)",
			input:   "loza://[::1]/demo",
			valid:   true,
			host:    "::1",
			port:    9308,
			project: "demo",
			tls:     false,
			baseURL: "http://[::1]:9308",
		},
		{
			name:    "tls=auto keeps localhost default",
			input:   "loza://localhost/demo?tls=auto",
			valid:   true,
			host:    "localhost",
			port:    9308,
			project: "demo",
			tls:     false,
			baseURL: "http://localhost:9308",
		},
		{
			name:    "tls=auto keeps remote default",
			input:   "loza://collector.example.com/demo?tls=auto",
			valid:   true,
			host:    "collector.example.com",
			port:    443,
			project: "demo",
			tls:     true,
			baseURL: "https://collector.example.com:443",
		},
		{
			name:    "explicit tls=true on localhost",
			input:   "loza://localhost:8443/demo?tls=true",
			valid:   true,
			host:    "localhost",
			port:    8443,
			project: "demo",
			tls:     true,
			baseURL: "https://localhost:8443",
		},
		{
			name:      "explicit port 4318 with otlp",
			input:     "loza://collector.example.com:4318/backend?env=staging&service=auth&transport=otlp",
			valid:     true,
			host:      "collector.example.com",
			port:      4318,
			project:   "backend",
			env:       "staging",
			service:   "auth",
			tls:       true,
			transport: "otlp",
			baseURL:   "https://collector.example.com:4318",
		},
		{
			name:      "credentialed DSN with percent-encoded password",
			input:     "loza://key-id:s%40cret%3Avalue%2Fpart@collector.example.com/demo",
			valid:     true,
			username:  "key-id",
			password:  "s@cret:value/part",
			host:      "collector.example.com",
			port:      443,
			project:   "demo",
			env:       "default",
			tls:       true,
			transport: "http",
			baseURL:   "https://collector.example.com:443",
		},
		{
			name:      "credentialed DSN with percent-encoded username",
			input:     "loza://key%2Did:secret@collector.example.com/demo?env=prod",
			valid:     true,
			username:  "key-id",
			password:  "secret",
			host:      "collector.example.com",
			port:      443,
			project:   "demo",
			env:       "prod",
			tls:       true,
			transport: "http",
			baseURL:   "https://collector.example.com:443",
		},

		// ── Invalid credential cases ─────────────────────────────────────────
		{
			name:  "reject empty username",
			input: "loza://:secret@collector.example.com/demo",
			valid: false,
		},
		{
			name:  "reject empty password",
			input: "loza://key:@collector.example.com/demo",
			valid: false,
		},
		{
			name:  "reject userinfo without password",
			input: "loza://key@collector.example.com/demo",
			valid: false,
		},
		{
			name:  "reject username containing encoded colon",
			input: "loza://key%3Aid:secret@collector.example.com/demo",
			valid: false,
		},
		{
			name:  "reject username containing encoded whitespace",
			input: "loza://key%20id:secret@collector.example.com/demo",
			valid: false,
		},
		{
			name:  "reject unencoded reserved password character",
			input: "loza://key:secret:part@collector.example.com/demo",
			valid: false,
		},
		{
			name:  "reject malformed percent escape in credentials",
			input: "loza://key:secret%ZZ@collector.example.com/demo",
			valid: false,
		},

		// ── Invalid cases ────────────────────────────────────────────────────

		{
			name:  "reject empty string",
			input: "",
			valid: false,
		},
		{
			name:  "reject wrong scheme https",
			input: "https://collector.example.com/demo",
			valid: false,
		},
		{
			name:  "reject wrong scheme http",
			input: "http://collector.example.com/demo",
			valid: false,
		},
		{
			name:  "reject no host",
			input: "loza://",
			valid: false,
		},
		{
			name:  "reject triple-slash no host",
			input: "loza:///project",
			valid: false,
		},
		{
			name:  "reject no project",
			input: "loza://collector.example.com",
			valid: false,
		},
		{
			name:  "reject empty project",
			input: "loza://collector.example.com/",
			valid: false,
		},
		{
			name:  "reject invalid tls value",
			input: "loza://collector.example.com/demo?tls=maybe",
			valid: false,
		},
		{
			name:  "reject invalid transport value",
			input: "loza://collector.example.com/demo?transport=random",
			valid: false,
		},
		{
			name:  "reject port 0",
			input: "loza://collector.example.com:0/demo",
			valid: false,
		},
		{
			name:  "reject port above 65535",
			input: "loza://collector.example.com:99999/demo",
			valid: false,
		},
		{
			name:  "reject non-numeric port",
			input: "loza://collector.example.com:abc/demo",
			valid: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			dsn, err := Parse(tt.input)

			if !tt.valid {
				if err == nil {
					t.Fatalf("expected error for %q, got valid DSN: %+v", tt.input, dsn)
				}
				return
			}

			if err != nil {
				t.Fatalf("expected valid DSN for %q, got error: %v", tt.input, err)
			}

			if dsn.Scheme != "loza" {
				t.Errorf("Scheme = %q, want %q", dsn.Scheme, "loza")
			}
			if dsn.Host != tt.host {
				t.Errorf("Host = %q, want %q", dsn.Host, tt.host)
			}
			if dsn.Port != tt.port {
				t.Errorf("Port = %d, want %d", dsn.Port, tt.port)
			}
			if dsn.Project != tt.project {
				t.Errorf("Project = %q, want %q", dsn.Project, tt.project)
			}
			if dsn.Username != tt.username {
				t.Errorf("Username = %q, want %q", dsn.Username, tt.username)
			}
			if dsn.Password != tt.password {
				t.Errorf("Password = %q, want %q", dsn.Password, tt.password)
			}
			if dsn.TLS != tt.tls {
				t.Errorf("TLS = %v, want %v", dsn.TLS, tt.tls)
			}
			if tt.transport != "" && dsn.Transport != tt.transport {
				t.Errorf("Transport = %q, want %q", dsn.Transport, tt.transport)
			}
			if dsn.BaseURL != tt.baseURL {
				t.Errorf("BaseURL = %q, want %q", dsn.BaseURL, tt.baseURL)
			}
			if tt.eventsURL != "" && dsn.EventsURL != tt.eventsURL {
				t.Errorf("EventsURL = %q, want %q", dsn.EventsURL, tt.eventsURL)
			}
			if tt.batchURL != "" && dsn.BatchURL != tt.batchURL {
				t.Errorf("BatchURL = %q, want %q", dsn.BatchURL, tt.batchURL)
			}
			if tt.otlpURL != "" && dsn.OTLPURL != tt.otlpURL {
				t.Errorf("OTLPURL = %q, want %q", dsn.OTLPURL, tt.otlpURL)
			}
			if tt.tailWSURL != "" && dsn.TailWSURL != tt.tailWSURL {
				t.Errorf("TailWSURL = %q, want %q", dsn.TailWSURL, tt.tailWSURL)
			}

			// Env defaults to "default" when not specified
			if tt.env != "" && dsn.Env != tt.env {
				t.Errorf("Env = %q, want %q", dsn.Env, tt.env)
			}
			if tt.service != "" && dsn.Service != tt.service {
				t.Errorf("Service = %q, want %q", dsn.Service, tt.service)
			}
		})
	}
}

func TestParseCredentialErrorsDoNotEchoPassword(t *testing.T) {
	const password = "super-secret"
	_, err := Parse("loza://key:" + password + ":part@collector.example.com/demo")
	if err == nil {
		t.Fatal("expected an error for an unencoded reserved password character")
	}
	if strings.Contains(err.Error(), password) {
		t.Fatalf("credential error echoed password: %q", err)
	}
}

func TestParseStringRedactsCredentials(t *testing.T) {
	dsn, err := Parse("loza://key-id:s%40cret%3Avalue@collector.example.com/demo")
	if err != nil {
		t.Fatal(err)
	}
	for _, format := range []string{"%v", "%+v", "%#v"} {
		rendered := fmt.Sprintf(format, dsn)
		if strings.Contains(rendered, "s@cret:value") || strings.Contains(rendered, "key-id") {
			t.Fatalf("format %q exposed credentials: %s", format, rendered)
		}
	}
}

func TestParseEnvDefault(t *testing.T) {
	dsn, err := Parse("loza://localhost/demo?tls=false")
	if err != nil {
		t.Fatal(err)
	}
	if dsn.Env != "default" {
		t.Errorf("Env = %q, want %q", dsn.Env, "default")
	}
}

func TestParseSchemeAlwaysLoza(t *testing.T) {
	dsn, err := Parse("loza://localhost/demo?tls=false")
	if err != nil {
		t.Fatal(err)
	}
	if dsn.Scheme != "loza" {
		t.Errorf("Scheme = %q, want %q", dsn.Scheme, "loza")
	}
}
